### PR TITLE
REM-1208: Integrate our BatchInjector

### PR DESCRIPTION
**Pull request by Art17**
### Jira references
- Ticket (task) — https://remmeio.atlassian.net/browse/REM-1208
### Obligatory payment handler
- Pull request — https://github.com/Remmeauth/remme-core/pull/244/files#diff-08c2207d72f248b35fb25bfddd7ab152R64

### CODE REVIEW

File: validator/sawtooth_validator/journal/batch_injector.py [hunk 1]
```diff
@@ -89,5 +89,12 @@ def _create_injector(self, injector):
 
             return block_info_injector.BlockInfoInjector(
                 self._state_view_factory, self._signer)
+        elif injector == "obligatory_payment":
+            block_info_injector = importlib.import_module(
```
Comment (eugene-babichenko) [inline validator/sawtooth_validator/journal/batch_injector.py]: The wrong object is created here, not our own injector.

File: validator/sawtooth_validator/journal/obligatory_payment_injector.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        self._state_view_factory = state_view_factory
+        self._signer = signer
+
+    def create_batch(self):
+        payload = ObligatoryPaymentPayload().SerializeToString()
+        public_key = self._signer.get_public_key().as_hex()
+
+        block_signer_address = make_address_from_data(data=public_key)
+
+        INPUTS = OUTPUTS = [
```
Comment (eugene-babichenko) [inline validator/sawtooth_validator/journal/obligatory_payment_injector.py]: Addressing is entirely wrong, should be fixed.

File: validator/sawtooth_validator/journal/obligatory_payment_injector.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        #TODO: also for inputs and outputs need addresses of other masternodes in committee.
+
+        header = TransactionHeader(
+            signer_public_key=public_key,
+            family_name=FAMILY_NAME,
+            family_version=FAMILY_VERSIONS[0],
+            inputs=INPUTS,
+            outputs=OUTPUTS,
+            dependencies=[],
+            payload_sha512=hash512(payload).hexdigest(),
```
Comment (eugene-babichenko) [inline validator/sawtooth_validator/journal/obligatory_payment_injector.py]: This causes an error.

**Review by eugene-babichenko — CHANGES_REQUESTED** [review]
1. Transactions/batches should always have some random `nonce`, not the default value. This is the requirement, because, transaction bodies will often be the same, signed by the same signers and thus have same ids. Using nonces will resolve that.
2. Not really sure should our batch be injected in the beginning or in the end of the block.
3. Our injector should also reuse batch injector for `BlockInfo` so that we have a single batch injector to control all injections being done.

File: validator/sawtooth_validator/journal/remme_batch_injector.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+            payload_sha512=hash512(payload),
+            batcher_public_key=public_key,
+            nonce=uuid4().hex
+        ).SerializeToString()
+
+        transaction_signature = self._signer.sign(header)
+
+        transaction = Transaction(
+            header=header,
+            payload=payload,
```
Comment (m-bo-one) [inline validator/sawtooth_validator/journal/remme_batch_injector.py]: Why payload here as `ObligatoryPaymentPayload` proto, if it required to be `TransactionPayload` for correct transaction processing?

**Pull request merged**